Fix #9352: Add ability to guess tie attachment point for different noteheads

asattely opened the pull request:
Resolves: https://github.com/musescore/MuseScore/issues/9352

Noteheads are now taken into account when determining where to attach ties. For noteheads with cutout information, this is the average of the cutouts. Also, now that noteheads have a single attachment point, ties adjust for each other when they share that point:
![image](https://user-images.githubusercontent.com/89263931/140239807-62f45c5c-dd4d-4739-a134-a29086aa22a2.png)

For slash noteheads, which are special, the tie attachment point takes the stem anchor into account:
![image](https://user-images.githubusercontent.com/89263931/140239843-28fe6942-6418-4f48-9068-c437a2f08232.png)

For all other noteheads, the exact center is used.
![image](https://user-images.githubusercontent.com/89263931/140239884-28a13aa9-ef31-4013-be9d-4404552f29a7.png)

Jojo-Schmitz commented:
3 vtests show differences, all with ties, unsurprisingly. 3 look OK, 3 don:t:
before:
![image](https://user-images.githubusercontent.com/1786669/140394003-6e95ab11-3f2b-496c-a0b7-d86ac0cf7a27.png)
after:
![image](https://user-images.githubusercontent.com/1786669/140394047-99cdfdb3-e2ba-4a7f-9f2b-1c38b107004c.png)
The tie in the 2nd measure is too high
before:
![image](https://user-images.githubusercontent.com/1786669/140394162-a73528ed-a003-410f-8b33-7991b30ba1de.png)
after:
![image](https://user-images.githubusercontent.com/1786669/140394211-6a9d3d54-9c2d-4702-9aa8-7b190e236c66.png)
In top staff the 1st tie in the 1st measure is too high, in bottom staff the last ties in the 1sxt meaure seem to ba to high/low
but the worst, before:
![image](https://user-images.githubusercontent.com/1786669/140394604-851d8c3a-eee2-4692-8c82-16f75f7ad513.png)
after
![image](https://user-images.githubusercontent.com/1786669/140394633-d55028c1-e896-40b7-9a25-63402363fc3a.png)
 the bottom tie starts way too late. It started to late before already, but now even later

Jojo-Schmitz commented:
Actually there's one more, before:
![image](https://user-images.githubusercontent.com/1786669/140395493-6c90b8ac-9dd6-4814-8333-627251fd11b2.png)
After:
![image](https://user-images.githubusercontent.com/1786669/140395529-985ff4d2-0b83-4065-ae5c-784c0e5c04d3.png)
See the tie for the 1st grace to note and  2 gracenotes at the end, they are too far to the right

asattely commented:
Yup, I noticed those as well. Thanks for looking through and getting the screenshots, I'll definitely be taking another whack at this once I finish something I'm working on tomorrow.

asattely commented:
So, the extra height on some of the ties were due to the fact that the vtests still use Emmentaler rather than Leland, for the record. The vtests will almost certainly still fail because the ties will look different, but now the ties from gracenotes will be a lot nicer with a little special handling. The wholenotes one was a weird oversight that should be fixed now. Let's see what the vtests say on this one.

Jojo-Schmitz commented:
Looks better, but this here does not, before:
![image](https://user-images.githubusercontent.com/1786669/141078247-ee839e40-478a-4cd8-aa63-1662470d09a9.png)
after:
![image](https://user-images.githubusercontent.com/1786669/141078294-59deb2c9-f492-42ab-94bf-9bb71ece8277.png)
The 2st tie looks better, the other 3 don't, they go too far past the note they tie to

asattely commented:
Yep, very well aware. The problem is that headWidth (and, bafflingly, the bounding box as well) always returns the width of a regular-sized notehead instead of the grace-sized one, so I just need to figure out how to get the grace-sized head width and we're golden.

Jojo-Schmitz commented:
See (in 3.x) Format > Style > Sites > Grace note size. Defaults to 70%, so you'ds need to pick on that style setting

asattely commented:
The answer was `mag()` (or `magS()`) which holds information about the note's relative size

asattely commented:
Those are inside-style ties because they are going to or from a multinote chord. This is an issue in other inside-style tie situations as well, and in the future there will be a minimum tie width issue where that will be looked at in greater detail.

asattely commented:
Still messing around with different fonts and noteheads, there are so many combinations to go through..!

asattely commented:
Rebase finished. I was also requested to add a small left- or right-offset to outside tie endpoints, which I ended up making 1/8 sp

Jojo-Schmitz commented:
4 utests crash with a failed assertion:
ASSERT: "_anchor == Anchor::CHORD" in file .../src/engraving/libmscore/spanner.cpp, line 879

asattely commented:
Tests should be fixed now

Jojo-Schmitz commented:
Nope, exact same crash

asattely commented:
The weird thing is that these tests pass when I run them locally. I think I identified the problem, we'll see if the tests are good this time

AntonioBL commented:
The vtests are now better, except for "grace-5-1".
Before
![grace-5-1 ref](https://user-images.githubusercontent.com/2855332/142613764-3d3a0bfd-a363-432b-b17d-85945b7df569.png)
After
![grace-5-1](https://user-images.githubusercontent.com/2855332/142613781-803392b3-1c78-415c-a5f8-700a1584b73f.png)
First, second and third ties are better than the reference, but there is something strange in the vertical position of the tie between the two grace notes (i.e. the fourth tie in the file).

Ciao

Nick-Mazuk commented:
Does this work as expected on chords where every note has a different notehead?

asattely commented:
Fixed a little issue with forced-direction outside ties colliding with hooks, and also added another special case for slashed noteheads, which are different from slash noteheads for the record.
